Clarify "don't stick out" considerations (#382)
tlswg/draft-ietf-tls-esni: draft-ietf-tls-esni.md

if it is invalid. It otherwise ignores the extension and MUST NOT use the retry
keys.

> [[TODO: Provide a defense for the GREASE+resumption and GREASE+PSK

Reviewer: Let's revert this. We don't yet have consensus as to whether or not we want (or need) to mitigate this particular differentiator. (Please also file a separate issue focused specifically on this issue that we can use to track discussion!)

Reviewer: (#354 seems too generic to be useful for converging on consensus here, I think.)

Author: Done. I also added a reference to #384.

TLS handshake as much as possible. The most obvious difference is the extension
itself: as long as middleboxes ignore it, as required by {{!RFC8446}}, the rest
of the handshake is designed to look very much as usual.

The GREASE ECH protocol described in {{grease-ech}} provides a low-risk way to
evaluate the deployability of ECH. It is designed to mimic the real ECH protocol
({{real-ech}}) without changing the security properties of the handshake. The
underlying theory is that if GREASE ECH is deployable without triggering
udifferentiable behavior, and real ECH is indifferentiable from GREASE ECH, then
ECH should be deployable as well. Thus, our strategy for mitigating network
ossification has two parts:
1. Design a "cover" protocol (i.e., GREASE ECH) that meets some notion of
   deployability.
1. Design the "target" protocol (i.e., ECH) so that networks do not differentiate
   between it and the cover protocol.

This strategy is analogous to the methodology that lead to "Middlebox
Compatibility Mode" for TLS 1.3 described in {{!RFC8446}}.

In the first step, the deployability of GREASE ECH is established by deploying

> the protocol and measuring its impact. [[OPEN ISSUE: Run an experiment that

Reviewer 2: Is this an open issue in the sense that the data from that experiment needs to be in this draft? If not, I'm not sure it should be mentioned here.

Reviewer: I think the point is that we will first run an experiment -- before this document is published as an RFC -- and then update this text with the results. Perhaps this helps?

```suggestion
the protocol and measuring its impact. [[TODO: Document results from preliminary experiment that
```

Author: Yup, that's what I mean!

include the plaintext extensions offered by the client or server, plaintext
alert messages, the number of round trips over the network, the length of each
message, and so on. We will call these the protocol's "observable features". Our
goal is to design real ECH so that each execution of the protocol has the same
observable features as if GREASE ECH were being executed instead. For example,
if extensions are an observable feature, then regardless of whether they are
running the target protocol or the cover protocol, the client and server should
appear to negotiate the same set of extensions. That way the real protocol does
not trigger differentiable treatment that would not have been triggered by the
cover protocol.

We say that the target protocol is "indifferentiable" from cover protocol if,
when executed in the presence of the adversary, both protocols have the same set
of observable features. The strength of this model depends on how rich the set
of observable features is: it may include the transcript of the protocol's
execution itself, yielding a very strong attacker; or it may be as coarse
grained as the number of IP packets sent. The model's strength also depends on

> whether the attacker is passive or active (see {{goals}}); middleboxes have been
> known to exhibit active behavior, even without malicious intent [[TOOD: Add
> references]].

Reviewer: ```suggestion
whether the attacker is passive or active (see {{goals}}).
```

I'd suggest dropping this unless we can point to some evidence. Maybe @davidben has one?

Author: There are some amazing examples [here](https://mailarchive.ietf.org/arch/msg/tls/i9blmvG2BEPf1s1OJkenHknRw9c/):
> We found that Firepower middleboxes in "Decrypt - Resign" mode
terminate TLS connections, but do not send a compliant ClientHello:
They modify the original ClientHello to remove unknown ciphersuites,
EMS, and NPN, but incorrectly forward most other fields from the
original ClientHello, including unknown extensions (supported_versions
and key_shares), and the client random. This breaks TLS 1.3 servers.
Additionally, these devices forward the server random rather than
generating their own (which will break when deploying the TLS 1.3
anti-downgrade feature), and forward unknown signature algorithms
(which will break when deploying, e.g., Ed25519).

Reviewer: Can we just cite this email? That would be fine for me!

Author: Cited references in Section D.4 of RFC 8446.

Indifferentiability of ECH from GREASE ECH in a strong attack model is not
feasible for all implementations. Thus, this specification aims to provide a
baseline security level that most deployments can achieve easily, while
providing implementations enough flexibility to achieve stronger security where
possible. Minimally, real ECH is designed to be indifferentiable from GREASE ECH
for passive adversaries with following capabilities:
1. The attacker does not know the ECHConfigs used by the server.
1. The attacker keeps per-connection state only. In particular, it does not
   track endpoints across connections.
1. Observable features include: the code points of extensions negotiated in the
   clear, but not the payloads themselves; the length of messages; and the
   values of plaintext alert messages.

This leaves a variety of practical differentiators out-of-scope. including,
though not limited to, the following:
1. the value of the configuration identifier;
1. the value of the outer SNI;
1. use of the "pre_shared_key" extension in the ClientHelloOuter, which is
   permitted in GREASE ECH but not real ECH; [[TODO: Based on discussion in

>    #354, we intend to mitigate this differentiator.]]

Reviewer: This isn't clear to me. I'd just leave this as an open issue without a decision one way or the other.

```suggestion
   #<new issue ID>, this may be mitigated with a future change.]]
```

### Do Not Stick Out {#dont-stick-out}

As a means of reducing the impact of network ossification, {{?RFC8744}}
recommends SNI-protection mechanisms be designed in such a way that network
operators do not differentiate connections using the mechanism from connections
not using the mechanism. To that end, ECH is designed to resemble a standard
TLS handshake as much as possible. The most obvious difference is the extension
itself: as long as middleboxes ignore it, as required by {{!RFC8446}}, the rest
of the handshake is designed to look very much as usual.

The GREASE ECH protocol described in {{grease-ech}} provides a low-risk way to
evaluate the deployability of ECH. It is designed to mimic the real ECH protocol

> ({{real-ech}}) without changing the security properties of the handshake. The

Reviewer 3: ```suggestion
({{real-ech}}) without protecting the client hello.
```

Author: I think this suggestion makes the text less clear.

underlying theory is that if GREASE ECH is deployable without triggering
udifferentiable behavior, and real ECH is indifferentiable from GREASE ECH, then

> ECH should be deployable as well. Thus, our strategy for mitigating network

Reviewer 3: I'm not sure I think of this so much as a measurement thing but rather as an anti-ossification measure.

Author: The point I was trying to make about measurement is that we have no idea how deployable GREASE ECH is today.

> ossification has two parts:
> 1. Design a "cover" protocol (i.e., GREASE ECH) that meets some notion of
>    deployability.
> 1. Design the "target" protocol (i.e., ECH) so that networks do not
>    differentiate between it and the cover protocol.

Reviewer 3: ```suggestion
ossification is to make the use of GREASE ECH widespread, thus disincentivizing filtering of the ECH extension by passive attackers.
```

Author: Changed as follows:
> Thus, our strategy for mitigating network ossification is to deploy GREASE ECH widely enough to disincentivize    differential treatment of the real ECH protocol by the network.

Compatibility Mode" for TLS 1.3 described in {{!RFC8446}}.

In the first step, the deployability of GREASE ECH is established by deploying
the protocol and measuring its impact. [[TODO: Document the results of an
experiment that compares the error rate of GREASE ECH connections with vanilla
TLS 1.3.]]

From this experiment we can deduce the set of features of the TLS handshake that
are known to trigger differentiable treatment by the network. This set might
include the plaintext extensions offered by the client or server, plaintext
alert messages, the number of round trips over the network, the length of each
message, and so on. We will call these the protocol's "observable features". Our
goal is to design real ECH so that each execution of the protocol has the same
observable features as if GREASE ECH were being executed instead. For example,
if extensions are an observable feature, then regardless of whether they are
running the target protocol or the cover protocol, the client and server should
appear to negotiate the same set of extensions. That way the real protocol does
not trigger differentiable treatment that would not have been triggered by the

> cover protocol.

Reviewer 3: I would remove everything between "This strategy is..." and here.

Author: Done.

We say that the target protocol is "indifferentiable" from cover protocol if,
when executed in the presence of the adversary, both protocols have the same set
of observable features. The strength of this model depends on how rich the set
of observable features is: it may include the transcript of the protocol's
execution itself, yielding a very strong attacker; or it may be as coarse
grained as the number of IP packets sent. The model's strength also depends on
whether the attacker is passive or active (see {{goals}}); middleboxes have been
known to exhibit active behavior, even without malicious intent. (See references

> in Section D.4 of {{!RFC8446}} for examples of active middlebox behavior.)

Reviewer 3: I would also remove this paragraph.

Author: Done.
